Add character count on edit document page

alex-ju opened the pull request:
This PR adds a character count to the Title (65 chars) and Summary (160 chars) fields on the edit document page. The code for this PR is based on the upcoming [Character count component](https://github.com/alphagov/govuk-frontend/pull/959) in govuk-frontend.

[Trello card](https://trello.com/c/28Ol82Ky)

<img width="658" alt="screen shot 2018-08-21 at 16 21 52" src="https://user-images.githubusercontent.com/788096/44411406-59637800-a55e-11e8-9c6b-bf3b69d6622a.png">

alex-ju commented:
@kevindew yes, I rely much on the Github PR message instead of commit message. I should mirror these because different people have different ways of reviewing. Using a branch from npm is a good idea, unfortunately content-publisher doesn't yet consume frontend directly, but via govuk_publishing_components gem, but after updating it to consume govuk-frontend directly I'll take the branch approach.

benthorner commented:
@kevindew Should this have a small spin-off story for us to update the validations?

benthorner commented on app/views/documents/edit.html.erb:
```diff
@@ -55,6 +57,8 @@
         value: @document.summary,
         rows: 4,
         data: {
+          "module": "character-count",
```
It looks like this kind of technique doesn’t extend if we wanted to add more modules to this element - would it be better to attach it in a similar way as contextual guidance?

alex-ju replied:
Yes, that is true, this method allows only one module per container, but generally I would avoid initialising multiple scripts on the same component as they may hijack each other's events. One script per component is consistent with how JavaScript is attached to all the components that are coming from `govuk-frontend`. If there are scripts acting at a document level, as it was the case of the contextual guidance, it makes sense for that script to use a `data-[component-name]` attribute selector or a `.js-[component-name]` class selector as we don't want to overlap with a component model and we want to be able to identify which script manipulates/affects the component. This is one of those things that I should document and explain at the project level.

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,187 @@
+  var countAttribute = (this.options.maxwords) ? this.defaults.wordCountAttribute : this.defaults.characterCountAttribute
+
+  // Set the element limit
+  if ($module.getAttribute) {
```
Should we just quit early here instead of having another if later on?

alex-ju replied:
Yes, we can quit early here. Maybe with a console message to mention an attribute is missing. Will update.

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,187 @@
+  this.options = this.merge(options || {}, dataset)
+
+  // Determine the limit attribute (characters or words)
+  var countAttribute = (this.options.maxwords) ? this.defaults.wordCountAttribute : this.defaults.characterCountAttribute
```
It looks like we’re doing this on characters - why are we supporting words?

alex-ju replied:
It was part of the original script which allows both characters and words. I pulled this in 'as is' because this is how it will come from `govuk-frontend`. Can remove the maxwords for now.

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,187 @@
+}
+
+// Fills in default values
+CharacterCount.prototype.merge = function (obj) {
```
Would it be easier to just manually merge attributes? I think there are only two of them ;-)

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,187 @@
+  return obj
+}
+
+CharacterCount.prototype.getDataset = function (element) {
```
Again, would it be easier to just fish these out manually if there aren’t many?

benthorner replied:
Or perhaps extract this to somewhere more general as it looks like it could be generally useful :-)

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,187 @@
+  var length
+  if (options.maxwords) {
+    var tokens = text.match(/\S+/g) || [] // Matches consecutive non-whitespace chars
+    length = tokens.length
```
You can just return here rather than have the variable - would be a bit easier to follow then :-)

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,187 @@
+
+// Generate count message and bind it to the input
+// returns reference to the generated element
+CharacterCount.prototype.createCountMessage = function () {
```
Again, wouldn’t it be easier to just the character count as actual HTML like contextual guidance?

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,187 @@
+  if (!this.oldValue) this.oldValue = ''
+  if (this.$module.value !== this.oldValue) {
+    this.oldValue = this.$module.value
+    var boundUpdateCountMessage = this.updateCountMessage.bind(this)
```
Why do we need a variable here?

alex-ju replied:
We need to store and execute the bounded function so no matter how and when it is called (on which component's instance), is called with a particular this (in our case our component's instance).

benthorner replied:
Hmm...so is `this.updateCountMessage.bind(this)()` not equivalent?

alex-ju replied:
nope

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,187 @@
+  var remainingNumber = maxLength - currentLength
+
+  // Set threshold if presented in options
+  var threshold = 0
```
You could make this a ternary - it seems to be the pattern used in the code above.

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,187 @@
+  if (options.threshold) {
+    threshold = options.threshold
+  }
+  var thresholdValue = maxLength * threshold / 100
```
Er, where’s the magic 100 coming from?

alex-ju replied:
`threshold` is a percentage value

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,187 @@
+  }
+
+  // Generate and reference message
+  var boundCreateCountMessage = this.createCountMessage.bind(this)
```
How come we need a variable here?

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,187 @@
+    $module.setAttribute('data-maxlength', $module.maxLength)
+
+    // Bind event changes to the textarea
+    var boundChangeEvents = this.bindChangeEvents.bind(this)
```
How come we need a variable here?

alex-ju commented:
> So this seems like quite a lot of code to generate single line of text.

I wouldn't judge or measure the code by how much the users sees 😉 

It may look a bit complex because it deals with both scenarios (character and word count). This is already a cleaned-up version without the highlight functionality which in my opinion contains all the complexity.

Updated:
- Refactored some parts (i.e. return early when possible, use ternary operator consistently)
- Removed word count option as we don't need it right now

Regarding `CharacterCount.prototype.merge` and  `CharacterCount.prototype.getDataset ` they are generic utility functions, but I wouldn't extract them until they are needed in other components.

Tried to reply the other questions inline, please let me know if I missed any.

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,166 @@
+  this.options = this.merge(options || {}, dataset)
+
+  // Read the limit attribute
+  var countAttribute = 'data-maxlength'
```
Do we need this variable now?

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,166 @@
+  this.countMessage = boundCreateCountMessage()
+
+  // If there's a maximum length defined and the count message was successfully applied
+  if (this.maxLength && this.countMessage) {
```
Again, can we return early here instead of having the block?

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,166 @@
+
+  // Set threshold if presented in options
+  var thresholdPercent = options.threshold ? options.threshold : 0
+  var thresholdValue = maxLength * thresholdPercent / 100
```
I still don't understand what this threshold percent thing is for - how does it relate to the UI?

benthorner commented on app/javascript/components/character-count.js:
```diff
@@ -0,0 +1,166 @@
+  }
+
+  // Update message
+  var charVerb = 'remaining'
```
Do we need these variables?

alex-ju commented:
Closing this as I consider unproductive to extract and branch code from a component that will replace this one. Thanks @benthorner for reviewing! For now we can use a hint text to state the maximum limit and rely on server side for validation. Will bring the character count component from `govuk-frontend` when is ready.
